sdk/compiler/daml-lf-proto-decode/src/DA/Daml/LF/Proto3/DecodeV2.hs

```diff
@@ -735,17 +739,37 @@ decodeNumericLit (T.unpack -> str) = case readMaybe str of
     Nothing -> throwError $ ParseError $ "bad Numeric literal: " ++ show str
     Just n -> pure $ BENumeric n
 
+singletonIfLfFlat :: Foldable t => t a -> Decode (t a)
+singletonIfLfFlat xs = do
+  v <- asks version
+  if v `supports` featureFlatArchive && length xs /= 1
+    then error $ printf "multiple arguments disallowed since lf %s supports flat archives" $ show v
```

Comment:
Can you replace this with `throwError` of a structured error from `compiler/daml-lf-proto-util/src/DA/Daml/LF/Proto3/Error.hs`? Feel free to add a new variant if none of them fit.